### PR TITLE
Upgrade password reset flow

### PR DESCRIPTION
**Pull request by LarsSelbekk**
- fix(error-handler): don't reuse accessToken invalid code
- fix: remove duplicated bad format error code
- feat: allow creating db-doc with given ID
- feat: upgrade password reset flow

### CODE REVIEW

**Comment by LarsSelbekk** [comment]
[Sikkerhetshull: kan endre hvem som helst sitt passord](https://trello.com/c/xkpYdkXG/812-sikkerhetshull-kan-endre-hvem-som-helst-sitt-passord)

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    this.messenger ??= new Messenger();
+  }
+
+  override async before(
+    passwordResetRequest: unknown,
+  ): Promise<PendingPasswordReset> {
+    validatePasswordResetRequest(passwordResetRequest);
+    if (!passwordResetRequest) {
+      throw new BlError("passwordResetRequest is empty or undefined");
+    }
```
Comment (AdrianAndersen) [inline src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts:31]: remove pls

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    if (!passwordResetRequest.email || !isEmail(passwordResetRequest.email)) {
+      throw new BlError(`passwordResetRequest.email is not a valid email`);
+    }
```
Comment (AdrianAndersen) [inline src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts:35]: dis too

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    const user = await this.userHandler
+      .getByUsername(passwordResetRequest.email)
+      .catch((getUserError: BlError) => {
+        throw new BlError(
+          `username "${passwordResetRequest.email}" not found`,
+        ).add(getUserError);
+      });
```
Comment (AdrianAndersen) [inline src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts:43]: needs an error code

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    const id = this.seCrypto.random();
+    const token = this.seCrypto.random();
+    const salt = this.seCrypto.random();
+    const tokenHash = await this.seCrypto.hash(token, salt);
+
+    // We should really wait to send the email until the password reset has been successfully written to the
+    // database, but it would be poor security to save the unhashed token in the database, and we have no other way
+    // of passing information from before to after, so this is the lesser evil.
+    await this.messenger
+      .passwordReset(user.id, passwordResetRequest.email, `${id}:${token}`)
```
Comment (AdrianAndersen) [inline src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts:58]: Please concat the id and token inside this method

File: src/collections/pending-password-reset/operations/confirm-pending-password-reset.operation.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      BlCollectionName.PendingPasswordResets,
+      pendingPasswordResetSchema,
+    );
+    this.localLoginHandler ??= new LocalLoginHandler();
+    this.responseHandler ??= new SEResponseHandler();
+    this.seCrypto ??= new SeCrypto();
+  }
+
+  async run(blApiRequest: BlApiRequest): Promise<BlapiResponse> {
+    const [pendingPasswordResetId, candidateToken] =
```
Comment (AdrianAndersen) [inline src/collections/pending-password-reset/operations/confirm-pending-password-reset.operation.ts:35]: Can you use query params here?

**Review by AdrianAndersen — CHANGES_REQUESTED** [review]
Please make sure the create password reset always returns 200 to avoid username lookups

**Comment by AdrianAndersen** [comment]
I still get everything back from the http://localhost:1337/pendingpasswordresets endpoint when I create a password reset 🧑‍🏫

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts (hunk 1 above)
Comment (LarsSelbekk) [reply src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts]: Done

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts (hunk 2 above)
Comment (LarsSelbekk) [reply src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts]: Done

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts (hunk 3 above)
Comment (LarsSelbekk) [reply src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts]: Done

File: src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts (hunk 4 above)
Comment (LarsSelbekk) [reply src/collections/pending-password-reset/hooks/pending-password-reset-post.hook.ts]: N/A when using query params

File: src/collections/pending-password-reset/operations/confirm-pending-password-reset.operation.ts (hunk 5 above)
Comment (LarsSelbekk) [reply src/collections/pending-password-reset/operations/confirm-pending-password-reset.operation.ts]: Done

**Comment by LarsSelbekk** [comment]
> I still get everything back from the http://localhost:1337/pendingpasswordresets endpoint when I create a password reset 🧑‍🏫

Fixed

**Review by AdrianAndersen — APPROVED** [review]

**Pull request merged**